bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/profiles/ScriptProfile.java

```diff
@@ -0,0 +1,158 @@
+
+    public ScriptProfile(ProfileCallback callback, ProfileContext context, RuleManager ruleManager,
+            RuleRegistry ruleRegistry) {
+        this.callback = callback;
+        this.ruleManager = ruleManager;
+        this.ruleRegistry = ruleRegistry;
+
+        var config = context.getConfiguration();
+        onCommandFromItemParam = getParam(config, ON_COMMAND_FROM_ITEM_PARAM);
+        onStateUpdateFromHandlerParam = getParam(config, ON_STATE_UPDATE_FROM_HANDLER_PARAM);
+        onCommandFromHandlerParam = getParam(config, ON_COMMAND_FROM_HANDLER_PARAM);
+        onStateUpdateFromItemParam = getParam(config, ON_STATE_UPDATE_FROM_ITEM_PARAM);
+    }
+
+    @Override
+    public ProfileTypeUID getProfileTypeUID() {
+        return PROFILE_TYPE_UID;
+    }
+
+    @Override
+    public void onCommandFromItem(Command command) {
+        if (runScript(ON_COMMAND_FROM_ITEM_PARAM, onCommandFromItemParam, COMMAND_VAR, command)) {
+            callback.handleCommand(command);
+        }
+    }
+
+    @Override
+    public void onStateUpdateFromHandler(State state) {
+        if (runScript(ON_STATE_UPDATE_FROM_HANDLER_PARAM, onStateUpdateFromHandlerParam, STATE_VAR, state)) {
+            callback.sendUpdate(state);
+        }
+    }
+
+    @Override
+    public void onCommandFromHandler(Command command) {
+        if (runScript(ON_COMMAND_FROM_HANDLER_PARAM, onCommandFromHandlerParam, COMMAND_VAR, command)) {
+            callback.sendCommand(command);
+        }
+    }
+
+    @Override
+    public void onStateUpdateFromItem(State state) {
+        runScript(ON_STATE_UPDATE_FROM_ITEM_PARAM, onStateUpdateFromItemParam, STATE_VAR, state);
+    }
+
+    private @Nullable String getParam(Configuration config, String paramName) {
+        Object param = config.get(paramName);
+        if (param != null && !(param instanceof String)) {
+            logger.error("Parameter '{}' must be a String. Profile will be inactive.", paramName);
+            return null;
+        }
+
+        logger.debug("Profile configured with {}='{}'", paramName, param);
+
+        return (String) param;
+    }
+
+    private boolean runScript(String callbackName, @Nullable String scriptName, String variableName,
+            Type variableValue) {
+        if (scriptName == null) {
+            return true;
+        }
+
+        var rule = ruleRegistry.get(scriptName);
+        if (rule != null && !rule.getTags().contains(SCRIPT_TAG)) {
+            rule = null;
+        }
+        if (rule == null) {
+            logger.warn("Script '{}' is missing for callback {}. Taking no action.", scriptName, callbackName);
+            // It was properly configured, so don't run the default actions.
+            return false;
+        }
+        var context = new HashMap<String, Object>();
+        context.put(CALLBACK_VAR, callback);
+        context.put(variableName, variableValue);
+
+        ruleManager.runNow(scriptName, false, context);
```

Comment:
should this be encapsulated in a scheduler? this is currently called from the CommunicationManager, which is likely calling it from the event thread I believe. and AFAICT the rule engine doesn't wrap it in a thread (just a try/catch) - normally it happens from the thread pool of a trigger handler callback.

Interestingly, this means the RunRuleActionHandler is running other rules synchronously; not sure if that's expected or not.

Reply:
no, a scheduler is not necessary. CommunicationManager wraps all calls to profiles in an async SafeCaller.